extension/api.js

```diff
@@ -0,0 +1,148 @@
+                // on `*` to retrieve URL prefixes.
+                const pattern = template.split("?")[0] + "*";
+                patterns[pattern] = _extensionID;
+              });
+            });
+          } catch (err) {
+            Cu.reportError(err);
+          }
+
+          return patterns;
+        },
+
+        // `getRequestProperty()` returns the property identified by
+        // `propertyName` for a given `requestId` (= channel ID). This might
+        // not return a value if the channel does not exist anymore or there is
+        // no such property.
+        async getRequestProperty(requestId, propertyName) {
+          const wrapper = ChannelWrapper.getRegisteredChannel(
+            requestId,
+            context.extension.policy,
+            context.xulBrowser.frameLoader.remoteTab
+          );
+
+          try {
+            return wrapper?.channel
+              ?.QueryInterface(Ci.nsIPropertyBag)
+              ?.getProperty(propertyName);
+          } catch {
+            // It is possible the property does not exist (or everything
+            // miserably failed).
+            return null;
+          }
+        },
+
+        // `getAddonVersion()` returns the add-on version if it exists.
+        async getAddonVersion(addonId) {
+          const addon = await AddonManager.getAddonByID(addonId);
+
+          return addon && addon.version;
+        },
+
+        // `getPublicSuffix()` returns the public suffix/Effective TLD Service
+        // of the given URL.
+        // See: https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIEffectiveTLDService
+        async getPublicSuffix(url) {
+          try {
+            return eTLD.getBaseDomain(Services.io.newURI(url));
+          } catch (err) {
+            Cu.reportError(err);
+            return null;
+          }
+        },
+
+        // `onSearchEngineModified` is an event that occurs when the list of
+        // search engines has changed, e.g., a new engine has been added or an
+        // engine has been removed.
+        //
+        // See: https://searchfox.org/mozilla-central/source/toolkit/components/search/SearchUtils.jsm#145-152
+        onSearchEngineModified: new ExtensionCommon.EventManager({
+          context,
+          name: "addonsSearchExperiment.onSearchEngineModified",
+          register: (fire) => {
+            const onSearchEngineModifiedObserver = (
+              aSubject,
+              aTopic,
+              aData
+            ) => {
+              if (
+                aTopic !== SEARCH_TOPIC_ENGINE_MODIFIED ||
+                // We are only interested in these modified types.
+                !["engine-added", "engine-removed"].includes(aData)
+              ) {
+                return;
+              }
+
+              fire.async();
+            };
+
+            searchInitialized.then(() => {
+              Services.obs.addObserver(
```

Comment:
`addObserver` should be called synchronously, to make sure that it balances the `removeObserver call later.

If you want to exclude search engines from before the initialization, check `Services.search.isInitialized` before calling `fire.async()`.

Reply:
@Rob--W something like this?

```diff
diff --git a/extension/api.js b/extension/api.js
index 4886d1f..f030c8c 100644
--- a/extension/api.js
+++ b/extension/api.js
@@ -124,15 +124,15 @@ this.addonsSearchExperiment = class extends ExtensionAPI {
                 return;
               }
 
-              fire.async();
+              searchInitialized.then(() => {
+                fire.async();
+              });
             };
 
-            searchInitialized.then(() => {
-              Services.obs.addObserver(
-                onSearchEngineModifiedObserver,
-                SEARCH_TOPIC_ENGINE_MODIFIED
-              );
-            });
+            Services.obs.addObserver(
+              onSearchEngineModifiedObserver,
+              SEARCH_TOPIC_ENGINE_MODIFIED
+            );
 
             return () => {
               Services.obs.removeObserver(
```